engines/sci/sound/audio32.cpp

```diff
@@ -0,0 +1,920 @@
+}
+
+void Audio32::setBitDepth(uint8 depth) {
+	if (depth > _maxAllowedBitDepth) {
+		depth = _maxAllowedBitDepth;
+	}
+
+	_globalBitDepth = depth;
+}
+
+void Audio32::setNumOutputChannels(int16 numChannels) {
+	if (numChannels > _maxAllowedOutputChannels) {
+		numChannels = _maxAllowedOutputChannels;
+	}
+
+	_globalNumOutputChannels = numChannels;
+}
+
+#pragma mark -
+#pragma mark Playback
+
+uint16 Audio32::play(int16 channelIndex, const ResourceId resourceId, const bool autoPlay, const bool loop, const int16 volume, const reg_t soundNode, const bool monitor) {
+	Common::StackLock lock(_mutex);
+
+	if (channelIndex != kNoExistingChannel) {
+		AudioChannel &channel = getChannel(channelIndex);
+
+		if (channel.pausedAtTick) {
+			resume(channelIndex);
+			return MIN(65534, 1 + channel.stream->getLength().msecs() * 60 / 1000);
+		}
+
+		warning("Tried to resume channel %s that was not paused", channel.id.toString().c_str());
+		return MIN(65534, 1 + channel.stream->getLength().msecs() * 60 / 1000);
+	}
+
+	if (_numActiveChannels == _channels.size()) {
+		warning("Audio mixer is full when trying to play %s", resourceId.toString().c_str());
+		return 0;
+	}
+
+	// NOTE: SCI engine itself normally searches in this order:
+	//
+	// For Audio36:
+	//
+	// 1. First, request a FD using Audio36 name and use it as the
+	//    source FD for reading the audio resource data.
+	// 2a. If the returned FD is -1, or equals the audio map, or
+	//     equals the audio bundle, try to get the offset of the
+	//     data from the audio map, using the Audio36 name.
+	//
+	//     If the returned offset is -1, this is not a valid resource;
+	//     return 0. Otherwise, set the read offset for the FD to the
+	//     returned offset.
+	// 2b. Otherwise, use the FD as-is (it is a patch file), with zero
+	//     offset, and record it separately so it can be closed later.
+	//
+	// For plain audio:
+	//
+	// 1. First, request an Audio resource from the resource cache. If
+	//    one does not exist, make the same request for a Wave resource.
+	// 2a. If an audio resource was discovered, record its memory ID
+	//     and clear the streaming FD
+	// 2b. Otherwise, request an Audio FD. If one does not exist, make
+	//     the same request for a Wave FD. If neither exist, this is not
+	//     a valid resource; return 0. Otherwise, use the returned FD as
+	//     the streaming ID and set the memory ID to null.
+	//
+	// Once these steps are complete, the audio engine either has a file
+	// descriptor + offset that it can use to read streamed audio, or it
+	// has a memory ID that it can use to read cached audio.
+	//
+	// Here in ScummVM we just ask the resource manager to give us the
+	// resource and we get a seekable stream.
+
+	// TODO: This should be fixed to use streaming, which means
+	// fixing the resource manager to allow streaming, which means
+	// probably rewriting a bunch of the resource manager.
+	Resource *resource = _resMan->findResource(resourceId, true);
```

Comment:
For memory efficiency, especially considering low-memory targets. Audio data can be fairly large and there isn’t a good reason to load it all into memory instead of streaming, since the audio engine already treats all audio as streamed. Some games, like Phantasmagoria, have audio samples that are 12MiB—and there can be 5 or 8 streams at once, which means an even larger audio footprint if we’re mixing several sounds together (background music + ambient audio + speech + sfx, for instance).